### PR TITLE
Collapse mm

### PR DESCRIPTION
**Pull request by zander-prinsloo**
Replace `merge.data.table()` with `collapse::join()` when `match_type = "m:m"`

All checks pass (excl. vignettes due to existing failures in DEV). 

image

### CODE REVIEW

File: R/update_na_vals.R [hunk 1]
```diff
@@ -60,20 +60,24 @@ update_na_values <- function(dt,
...
-         eval(x.var) := mget(y.var)]
+      dt_1[dt_1[[reportvar]] == 4,
+           x.var] <- dt_1[dt_1[[reportvar]] == 4,
+                          y.var,
+                          with = FALSE]
+#
+      dt_1[dt_1[[reportvar]] == 5,
+           x.var] <- dt_1[dt_1[[reportvar]] == 5,
+                          y.var,
+                          with = FALSE]
```
Comment (randrescastaneda) [inline R/update_na_vals.R:71]: Hi @zander-prinsloo , I don't understand this change. It is now making a deep copy rather than modifying by reference. What is the purpose of this change?

File: R/joyn-merge.R [hunk 2]
```diff
@@ -231,6 +228,22 @@ joyn <- function(x,
...
+  # ensure input names can be restored
+  byexp <- grep(pattern = "==?",
+                x       = by,
+                value   = TRUE)
+  xbynames <- trimws(gsub("([^=]+)(\\s*==?\\s*)([^=]+)",
+                          "\\1",
+                          byexp))
+  ybynames <- trimws(gsub("([^=]+)(\\s*==?\\s*)([^=]+)",
+                          "\\3",
+                          byexp))
```
Comment (randrescastaneda) [inline R/joyn-merge.R:240]: what happened with function `fix_by_vars()`?

File: R/joyn-merge.R [hunk 3]
```diff
@@ -231,6 +228,22 @@ joyn <- function(x,
...
+  ynames  <- copy(names(y))
+
+  # maintain name that is bound to original inputs
+  x_original <- x
+  y_original <- y
+
   # If match type is m:m we need to convert to data.table
   if (match_type == "m:m") {
```
Comment (randrescastaneda) [inline R/joyn-merge.R:248]: This is not necessary anymore given that we are using `collapse::join` for `m:m`, right?

File: R/joyn-merge.R [hunk 4]
```diff
@@ -429,21 +439,49 @@ joyn <- function(x,
...
+
+    data.table::setnames(y_original,
+                         old = knames,
+                         new = ybynames)
+
+    if (all(names(y_original) %in% ynames)) {
+      colorderv(y_original,
+                neworder = ynames)
+    }
+  }
```
Comment (randrescastaneda) [inline R/joyn-merge.R:485]: When match_type is `m:m` this never hold because of the `as.data.table()` above, right? Is this by intention?

**Review by randrescastaneda — COMMENTED** [review]
Hi @zander-prinsloo , 

Thank you so much for the improvement. There are a few things that are still not clear to me. Let's talk about them before merging. 

Thanks.

File: R/update_na_vals.R (hunk 1 above)
Comment (zander-prinsloo) [reply R/update_na_vals.R:71]: @randrescastaneda , previously it was making a deep copy already using `copy()` and then modifying by reference after that. But you are right, this is not a sufficient improvement on what was done before. Let me propose a modification.

File: R/joyn-merge.R [hunk 5]
```diff
@@ -231,16 +228,60 @@ joyn <- function(x,
...
+  x_original <- x
+  y_original <- y
 
   ## Modify BY when is expression ---------
   fixby  <- check_by_vars(by, x, y)
   by     <- fixby$by
 
+  # Change names back on exit
+  # Change names back for inputs------------------------------
+  on.exit(
```
Comment (randrescastaneda) [inline R/joyn-merge.R:253]: I really like this idea. excellent.

File: R/dplyr-joins.R [hunk 6]
```diff
@@ -520,6 +595,19 @@ inner_join <- function(
   dropreport   <- args_check$dropreport
 
   # Column names -----------------------------------
+  byexp <- grep(pattern = "==?",
```
Comment (randrescastaneda) [inline R/dplyr-joins.R:598]: I think these lines of code should be converted to a function that can be called everywhere. Right now, the same code is repeated in multiple places.

**Comment by zander-prinsloo** [comment]
@randrescastaneda Repetitive code made into a function. Merged with `DEV` and conflicts resolved. Please merge if you're happy. 

Note seen below was introduced by masking functions, specifically `unmask_joyn_fun_ns()`

image

**Review by randrescastaneda — APPROVED** [review]
